### PR TITLE
Phlin/edge agent get twin fix

### PR DESCRIPTION
**Pull request by philipktlin**
1. When Edge agent calling GetTwinAsync method throws exception after retry policy, if the exception is not timeout exception it will trigger to close module client and retry again.
2. Remove cloud connection in connection manager when remove device connection for deployment scenario that a module is removed and added it back in a short period of time.  This will prevent cloud connection is in invalid state.
3. Upgrade Microsoft.Azure.Devices.Client NuGet package from 1.23.1 to 1.23.2, which contains AMQP Safely close fix.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs [hunk 1]
```diff
@@ -210,10 +210,21 @@ async Task<Twin> GetTwinFunc()
             catch (Exception e)
             {
                 Events.ErrorGettingTwin(e);
+
+                if (!forceNewModuleClient && this.RetryWithNewModuleClient(e))
+                {
+                    return await this.GetTwinFromIoTHub(true);
```
Comment (varunpuranik) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:216]: If this throws, will the exception bubble up? I beleive we don't want that.

Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:216]: since this method already caught everything, it will not have any exception thrown from this method.  When exception happens, it will catch and return Option.None<Twin>().

Comment (varunpuranik) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:216]: ah yes

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs [hunk 2]
```diff
@@ -210,10 +210,21 @@ async Task<Twin> GetTwinFunc()
...
+                }
+
                 return Option.None<Twin>();
             }
         }
 
+        bool RetryWithNewModuleClient(Exception e)
```
Comment (varunpuranik) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:223]: Nit: remove this method, not very useful

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs [hunk 3]
```diff
@@ -184,14 +184,14 @@ async Task RefreshTwinAsync()
...
+        async Task<Option<Twin>> GetTwinFromIoTHub(bool forceNewModuleClient = false)
         {
             try
             {
                 async Task<Twin> GetTwinFunc()
                 {
-                    Events.GettingModuleClient();
-                    IModuleClient moduleClient = await this.moduleConnection.GetOrCreateModuleClient();
+                    Events.GettingModuleClient(forceNewModuleClient);
+                    IModuleClient moduleClient = await this.moduleConnection.GetOrCreateModuleClient(forceNewModuleClient);
```
Comment (varunpuranik) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:194]: This will cause the module client to be created every time in the retry operation (up to 3 times). 
We should do it only once. Maybe it may be easier to close the moduleClient in the code here and let GetOrCreateModuleClient create a new one.

Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:194]: If we come to this point, it means something happens to the module client other than connection timeout.  If that is the case, I expect it should be succeeded in 1st call and no retry should be needed.  If retry really happens, it is no harmful to re-create a module client.  I think this is more simple, isn't it?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs (hunk 2 above)
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:223]: Sure. I will do inline.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs [hunk 4]
```diff
@@ -210,6 +212,21 @@ async Task<Twin> GetTwinFunc()
             catch (Exception e)
             {
                 Events.ErrorGettingTwin(e);
+
+                if (!retrying && moduleClient != null && !(e is TimeoutException))
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:216]: Yes.  you are right.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs [hunk 5]
```diff
@@ -75,7 +75,7 @@ await currentDeviceConnection
         public Task RemoveDeviceConnection(string id)
         {
             return this.devices.TryGetValue(Preconditions.CheckNonWhiteSpace(id, nameof(id)), out ConnectedDevice device)
-                ? this.RemoveDeviceConnection(device, false)
+                ? this.RemoveDeviceConnection(device, true)
```
[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: will do.

[the remark this answers is not in the document]
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: Can we separate the 2 fixes out?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/Microsoft.Azure.Devices.Edge.Hub.Core.csproj [hunk 6]
```diff
@@ -30,7 +30,7 @@
   <ItemGroup>
     <PackageReference Include="App.Metrics" Version="3.0.0" />
     <PackageReference Include="JetBrains.Annotations" Version="2018.3.0" />
-    <PackageReference Include="Microsoft.Azure.Devices.Client" Version="1.23.1" />
+    <PackageReference Include="Microsoft.Azure.Devices.Client" Version="1.23.2" />
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/Microsoft.Azure.Devices.Edge.Hub.Core.csproj:33]: Let's also separate out this into another PR as well.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs (hunk 5 above)
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: Let's remove the flag altogether..

[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: Checked with Dylan. His run for connectivity test shows negative due to another setting issue.

[the remark this answers is not in the document]
Comment (philipktlin) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: will split it.

Comment (philipktlin) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: will do in another PR.

Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: Actually, let's add a setting (env var) to EdgeHub close the cloud connection when device is gone or not. Default will be to close the cloud connection, but the flag lets you turn it off.

Comment (philipktlin) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs:78]: 1. In what case we would like to keep the cloud connection even the module is removed from deployment?
2. There are totally 5 usage references and 2 of them passed false in 2nd argument.  One is this usage.  @varunpuranik  I would like to clarify, does env var setting to control all 5 usages or only those usages currently passing false as 2nd argument?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/Microsoft.Azure.Devices.Edge.Hub.Core.csproj (hunk 6 above)
Comment (philipktlin) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/Microsoft.Azure.Devices.Edge.Hub.Core.csproj:33]: ok.

File: test/Microsoft.Azure.Devices.Edge.Test.Common/Microsoft.Azure.Devices.Edge.Test.Common.csproj [hunk 7]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (philipktlin) [inline test/Microsoft.Azure.Devices.Edge.Test.Common/Microsoft.Azure.Devices.Edge.Test.Common.csproj:1]: Tried to manually revert this, but no difference.  Don't know why it shows up.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs [hunk 8]
```diff
@@ -354,9 +372,9 @@ public static void ErrorUpdatingReportedProperties(Exception ex)
                 Log.LogDebug((int)EventIds.ErrorUpdatingReportedProperties, ex, "Error updating reported properties in IoT Hub");
             }
 
-            public static void GettingModuleClient()
+            public static void GettingModuleClient(bool retrying)
             {
-                Log.LogDebug((int)EventIds.GettingModuleClient, "Getting module client to refresh the twin");
+                Log.LogDebug((int)EventIds.GettingModuleClient, $"Getting module client to refresh the twin with retrying set to {retrying}");
```
Comment (varunpuranik) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:377]: This is probably not useful anymore? Maybe we should remove

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleConnection.cs [hunk 9]
```diff
@@ -134,7 +134,7 @@ enum EventIds
                 ReceivedMethodCallback,
                 DisposingModuleConnection,
                 DisposedModuleConnection,
-                ErrorDisposingModuleConnection
+                ErrorDisposingModuleConnection,
```
Comment (varunpuranik) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleConnection.cs:137]: ? revert?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs (hunk 8 above)
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs:377]: it is good to display in the log to indicate that this call is retrying or not.   I think we should keep this.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleConnection.cs (hunk 9 above)
Comment (philipktlin) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleConnection.cs:137]: good catch.
